src/Raven.Server/Documents/RetiredAttachmentsStorage.cs

```diff
@@ -0,0 +1,482 @@
+        size = size - sizeOfIdentifier - 1; // skip identifier and record separator
+        byte* pp = ExtractHashFromAttachmentIdInternal(p, size);
+        var d1 = Slice.External(context.Allocator, pp, AttachmentsStorage.AttachmentHashSize, out hashSlice);
+
+        return d1;
+    }
+
+    internal static unsafe byte* ExtractHashFromAttachmentIdInternal(byte* p, int size)
+    {
+        AttachmentsStorage.ExtractDocIdAndAttachmentNameFromTombstone(p, size, out _, out _, out _, out int attachmentHashIndex);
+        var pp = p + attachmentHashIndex;
+        return pp;
+    }
+
+    protected override unsafe void ProcessDocument(DocumentsOperationContext context, Slice lowerId, string id, DateTime currentTime)
+    {
+        var sizeOfIdentifier = AttachmentsStorage.AttachmentKey.GetSizeOfDocId(new ReadOnlySpan<byte>(lowerId.Content.Ptr, lowerId.Size));
+        using (var lowerDocId = _documentInfoHelper.GetDocumentId(lowerId, sizeOfIdentifier + 1))
+        {
+            if (lowerDocId == null)
+            {
+                throw new InvalidOperationException($"Couldn't retire the attachment. Retired attachment key is '{lowerId}'.");
+            }
+
+            using (var doc = Database.DocumentsStorage.Get(context, lowerDocId, DocumentFields.Data | DocumentFields.Id, throwOnConflict: true))
+            {
+                if (doc == null || doc.TryGetMetadata(out var metadata) == false)
+                    return;
+
+                if (metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachments) == false)
+                    return;
+
+                using var attachmentKeyDisposable = ExtractAttachmentKeySliceFromRetiredAttachmentKey(context, lowerId, out Slice attachmentKeySlice);
+                var nameByKey = Database.DocumentsStorage.AttachmentsStorage.GetAttachmentNameByKey(context, attachmentKeySlice);
+                if (nameByKey == null)
+                    return;
+
+                for (var i = 0; i < attachments.Length; i++)
+                {
+                    var attachmentInMetadata = (BlittableJsonReaderObject)attachments[i];
+                    if (attachmentInMetadata.TryGet(nameof(AttachmentName.Name), out string name) == false)
+                        continue;
+                    if (attachmentInMetadata.TryGet(nameof(AttachmentName.RetireParameters), out BlittableJsonReaderObject retireParamsObject) == false)
+                        continue;
+                    if (retireParamsObject.TryGet(nameof(RetireAttachmentParameters.Identifier), out LazyStringValue identifierFromMetadata) == false)
+                        continue;
+                    if (identifierFromMetadata != id)
+                        continue;
+
+                    if (name == nameByKey)
+                    {
+                        if (HasPassed(retireParamsObject, currentTime, MetadataPropertyName) == false)
```

Comment:
what if I have multiple attachments that needs to be retired at different times? Aren't you returning early?

What does `HasPassed` means - why would we invoke this call?